File: bench/bench_struct.py
import time

class Vec2:
    def __init__(self, x, y):
        self.x = x
        self.y = y
    def mag_sq(self):
        return self.x * self.x + self.y * self.y

start = time.time()
total = 0.0
for i in range(50000):
    v = Vec2(float(i), float(i + 1))
    total += v.mag_sq()
elapsed = time.time() - start
print(int(total))
print(f"time: {elapsed:.6f}s")

File: bench/bench_struct.cpp
#include <stdio.h>
#include <time.h>

struct Vec2 {
    double x, y;
    double mag_sq() { return x*x + y*y; }
};

int main() {
    struct timespec start, end;
    clock_gettime(CLOCK_MONOTONIC, &start);
    double total = 0.0;
    for (int i = 0; i < 50000; i++) {
        Vec2 v = {(double)i, (double)(i+1)};
        total += v.mag_sq();
    }
    clock_gettime(CLOCK_MONOTONIC, &end);
    double elapsed = (end.tv_sec - start.tv_sec) + (end.tv_nsec - start.tv_nsec) / 1e9;
    printf("%lld\n", (long long)total);
    printf("time: %.6fs\n", elapsed);
    return 0;
}
